File: src/cxx-dev/LiENa/lienaRealTime/lienaEncodingTask.cpp
#include "lienaEncodingTask.h"


/**
 * @brief igtEncodingTask::igtEncodingTask
 * @param dataware
 * @param outputQueueManager
 * @param datagramAnalyser
 */
lienaEncodingTask::lienaEncodingTask(lienaOutputQueue *outputQueue, lienaGlobal *globalParameter, bool motivate, uint32_t targetDeviceId){
    this->flag = true;
    this->rtPeriod = 20;

    this->standby = true;

    this->motivate = motivate;
    this->targetDeviceId = targetDeviceId;

    this->msgQ = nullptr;
    this->outputQueue = outputQueue;
    this->globalParameter = globalParameter;
    this->encoder = new lienaDatagramEncoder(globalParameter);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::~lienaEncodingTask
//!
lienaEncodingTask::~lienaEncodingTask(){
    delete this->encoder;
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendDisengagementCommitMessage
//!
void lienaEncodingTask::sendDisengagementCommitMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask::sendDisengagementCommitMessage";
    }

    uint64_t message_id = LIENA_SESSION_MANAGEMENT_DISENGAGEMENTCOMMIT_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t  timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaDisengagementCommitMessage *disengagementCommitMessage = new lienaDisengagementCommitMessage(message_id, targetDeviceId,originId, timeStamps, 6);

    lienaDatagram *datagram = encoder->transformDisengagementCommitMessageToIgtDatagram(disengagementCommitMessage);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendDisengagementmessage
//!
void lienaEncodingTask::sendDisengagementMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask::sendDisengagementmessage :";
    }

    uint64_t message_id = LIENA_SESSION_MANAGEMENT_DISENGAGEMENT_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaDisengagementMessage *disengagementMessage = new lienaDisengagementMessage(message_id, targetDeviceId, originId, timeStamps, 6);

    lienaDatagram *datagram = encoder->transformDisengagementMessageToIgtDatagram(disengagementMessage);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendChannelClosedMessage
//!
void lienaEncodingTask::sendChannelClosedMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask::sendChannelClosedMessage";
    }

    uint64_t message_id = LIENA_SESSION_MANAGEMENT_CHANNELCLOSED_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaChannelClosedMessage *channelClosedMessage = new lienaChannelClosedMessage(message_id, targetDeviceId, originId, timeStamps, 6);

    lienaDatagram *datagram = encoder->transformchannelClosedMessageToIgtDatagram(channelClosedMessage);
    this->outputQueue->append(datagram);
}

//! -------------------------------------------------------------------------------------------------------------------------------------------------
//!
//! \brief igtClient::sendHandShakeCommitMessage
//!
void lienaEncodingTask::sendHandShakeCommitMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask::sendHandShakeCommitMessage";
    }

    uint64_t message_id = LIENA_SESSION_MANAGEMENT_HANDSHAKECOMMIT_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaHandShakeCommitMessage *handShakeCommitMessage = new lienaHandShakeCommitMessage(message_id, targetDeviceId, originId, timeStamps, 6);

    lienaDatagram *datagram = encoder->transformHandShakeCommitMessageToIgtDatagram(handShakeCommitMessage);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendReHandshakeMessage
//!
void lienaEncodingTask::sendReHandshakeMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask | sendReHandshakeMessage";
    }
    uint64_t message_id = LIENA_SESSION_MANAGEMENT_REHANDSHAKE_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaReHandshakeMessage *reHandShakeMessage = new lienaReHandshakeMessage(message_id, targetDeviceId, originId, timeStamps, 6);
    lienaDatagram *datagram = encoder->transformReHandShakeMessageToIgtDatagram(reHandShakeMessage);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendReHandshakeCommitMessage
//!
void lienaEncodingTask::sendReHandshakeCommitMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask | sendReHandshakeCommitMessage";
    }
    uint64_t message_id = LIENA_SESSION_MANAGEMENT_REHANDSHAKECOMMIT_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaReHandshakeCommitMessage *reHandShakeCommitMessage = new lienaReHandshakeCommitMessage(message_id, targetDeviceId , originId, timeStamps, 6);
    lienaDatagram *datagram = encoder->transformReHandShakeCommitMessageToIgtDatagram(reHandShakeCommitMessage);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendChannelReOpened
//!
void lienaEncodingTask::sendChannelRepairedMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask | sendChannelReOpened";
    }
    uint64_t message_id = LIENA_SESSION_MANAGEMENT_REOPENED_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaChannelReOpened *channelReOpenedMessage = new lienaChannelReOpened(message_id, targetDeviceId,originId,  timeStamps, 6);
    lienaDatagram *datagram = encoder->transformChannelReOpenedMessageToIgtDatagram(channelReOpenedMessage);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::sendChannelOpenedMessage
//!
void lienaEncodingTask::sendChannelOpenedMessage(){
    if(DEBUG){
        qDebug()<<"lienaEncodingTask::sendChannelOpenedMessage";
    }
    uint64_t message_id = LIENA_SESSION_MANAGEMENT_CHANNELOPENNED_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaChannelOpenedMessage *channelOpenedMsg = new lienaChannelOpenedMessage(message_id, targetDeviceId, originId, timeStamps, 6);
    lienaDatagram *datagram = encoder->transformChannelOpenedMessageToIgtDatagram(channelOpenedMsg);
    this->outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaTransmissionTask::sendHandShakeMessage
//!
void lienaEncodingTask::sendHandShakeMessage(QString addr){
    if(DEBUG){
        qDebug()<<"lienaTransmissionTask::sendHandShakeMessage";
    }

    uint64_t message_id = LIENA_SESSION_MANAGEMENT_HANDSHAKE_MESSAGE;
    uint32_t targetDeviceId = this->targetDeviceId;
    uint64_t timeStamps = this->globalParameter->getTimestamps();
    uint32_t originId = this->globalParameter->getLocalDeviceId();

    lienaHandShakeMessage *handShakeMessage = new lienaHandShakeMessage(message_id, targetDeviceId, originId, timeStamps, 6, addr, 10704);
    lienaDatagram *datagram = this->encoder->encode_handshake_message(handShakeMessage);
    outputQueue->append(datagram);
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::launch
//!
void lienaEncodingTask::launch(){
    this->standby = false;
    this->start();
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::enable
//!
void lienaEncodingTask::enable(){
    this->standby = false;
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::freeze
//!
void lienaEncodingTask::freeze(){
    this->standby = true;
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief igtEncodingTask::setPatientHandling
//! \param patientHandling
//!
void lienaEncodingTask::setOutputMessageQueue(lienaMessageQueue* msgQ){
    this->msgQ = msgQ;
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief igtEncodingTask::run
//!
void lienaEncodingTask::run(){
    while (flag) {

        if(standby){
            this->msleep(1000);
            continue;
        }
        if (this->msgQ != nullptr){
            if(!this->msgQ->isEmpty()){
                lienaCustomizedMessage *msg = this->msgQ->pop_front();



                //transformation(msg);


                outputQueue->append(this->encoder->encode_customized_message(msg));
            }
        }

        msleep(this->rtPeriod);
    }
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::setPeriod
//! \param rtPeriod
//!
void lienaEncodingTask::setPeriod(uint32_t rtPeriod){
    this->rtPeriod = rtPeriod;
}

//! ---------------------------------------------------------------------------------------------------------------
//!
//! \brief lienaEncodingTask::stop
//!
void lienaEncodingTask::stop(){
    this->flag = false;
}

File: src/py-dev/LiENa/LiENaRealTimeTask/LienaEncodingTask.py
# -*- coding: utf-8 -*-
import threading
import time
from LiENaBasic.lienaDefinition import *
from LiENaStructure.LiENaDatagram.LienaEncoder import LienaEncoder
from LiENaStructure.LiENaMessage.LienaHandShakeMessage import LienaHandShakeMessage
from LiENaStructure.LiENaMessage.LienaChannelOpenedMessage import LienaChannelOpenedMessage
from LiENaStructure.LiENaMessage.LienaHandShakeCommitMessage import LienaHandShakeCommitMessage
from LiENaStructure.LiENaMessage.LienaDisengagementCommitMessage import LienaDisengagementCommitMessage
from LiENaStructure.LiENaMessage.LienaDisengagementMessage import LienaDisengagementMessage
from LiENaStructure.LiENaMessage.LienaChannelClosedMessage import LienaChannelClosedMessage
from LiENaStructure.LiENaMessage.LienaReHandshakeCommitMessage import LienaReHandshakeCommitMessage
from LiENaStructure.LiENaMessage.LienaChannelReOpenedMessage import LienaChannelReOpenedMessage
from LiENaStructure.LiENaMessage.lienaReHandshakeMessage import LienaReHandshakeMessage


class LienaEncodingTask:
    def __init__(self, output_queue, global_parameter, motivate, target_device_id):
        self.output_queue = output_queue
        self.global_parameter = global_parameter
        self.motivate = motivate
        self.targetDeviceId = target_device_id
        self.encoder = LienaEncoder(self.global_parameter)

        self.stand_by = False

        self.outputMessageCache = None
        self.rtPeriod = 0.05

        self.flag = True
        self.encodingThread = threading.Thread(None, self.encode)

        self.msgQueue = None

    def set_output_msg_queue(self, outputMegQue):
        self.msgQueue = outputMegQue

    def enable(self):
        self.stand_by = False

    def freeze(self):
        self.stand_by = True

    def send_channel_opened_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_channel_opened_message")

        message_id = self.global_parameter.get_local_device_id() * (2 ** 32) + LIENA_SESSION_MANAGEMENT_CHANNEL_OPENED_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)

        channel_opened_message = LienaChannelOpenedMessage(message_id, target_device_id, origin_id,  timestamps, 6)
        datagram = self.encoder.encode(channel_opened_message)
        self.output_queue.append(datagram)

    def send_disengagement_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_disengagement_message")

        message_id =  LIENA_SESSION_MANAGEMENT_DISENGAGEMENT_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        disengagement_message = LienaDisengagementMessage(message_id, target_device_id, origin_id, timestamps, 6)
        
        datagram = self.encoder.encode(disengagement_message)
        self.output_queue.append(datagram)

    def send_channel_repaired_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_channel_repaired_message")

        message_id = LIENA_SESSION_MANAGEMENT_CHANNEL_REOPENED_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        disengagement_message = LienaChannelReOpenedMessage(message_id, target_device_id, origin_id, timestamps, 6)

        datagram = self.encoder.encode(disengagement_message)
        self.output_queue.append(datagram)


    def send_rehandshake_commit_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_rehandshake_commit_message")

        message_id = LIENA_SESSION_MANAGEMENT_REHANDSHAKECOMMIT_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        message = LienaReHandshakeCommitMessage(message_id, target_device_id, origin_id, timestamps, 6)

        datagram = self.encoder.encode(message)
        self.output_queue.append(datagram)

    def send_disengagement_commit_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_disengagement_commit_message")

        message_id = LIENA_SESSION_MANAGEMENT_DISENGAGEMENTCOMMIT_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        message = LienaDisengagementCommitMessage(message_id, target_device_id, origin_id, timestamps, 6)

        datagram = self.encoder.encode(message)
        self.output_queue.append(datagram)

    def send_back_network_quality_message(self, msg):
        datagram = self.encoder.encode(msg)
        self.output_queue.append(datagram)

    def send_rehandshake_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_rehandshake_message")

        message_id = LIENA_SESSION_MANAGEMENT_REHANDSHAKE_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        message = LienaReHandshakeMessage(message_id, target_device_id, origin_id, timestamps, 6)

        datagram = self.encoder.encode(message)
        self.output_queue.append(datagram)

    def send_channel_closed_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_channel_closed_message")

        message_id = LIENA_SESSION_MANAGEMENT_CHANNEL_CLOSED_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        channel_closed_message = LienaChannelClosedMessage(message_id, target_device_id, origin_id,  timestamps, 6)

        datagram = self.encoder.encode(channel_closed_message)
        self.output_queue.append(datagram)

    def send_handshake_commit_message(self):
        if DEBUG:
            print("LienaEncodingTask | send_handshake_commit_message")
        message_id = LIENA_SESSION_MANAGEMENT_HANDSHAKE_COMMIT_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        handshake_commit_message = LienaHandShakeCommitMessage(message_id, target_device_id, origin_id, timestamps, 6)

        datagram = self.encoder.encode(handshake_commit_message)
        self.output_queue.append(datagram)

    def send_handshake_message(self, addr):
        if DEBUG:
            print("LienaEncodingTask | new_send_handshake_message", addr)

        message_id = LIENA_SESSION_MANAGEMENT_HANDSHAKE_MESSAGE
        target_device_id = self.targetDeviceId
        origin_id = self.global_parameter.get_local_device_id()
        timestamps = round((time.time() % 86400) * 1000000)
        handshake_message = LienaHandShakeMessage(message_id, target_device_id, origin_id, timestamps, 6, addr, 10704)

        datagram = self.encoder.encode(handshake_message)
        self.output_queue.append(datagram)

    def launch(self):
        self.encodingThread.start()

    def set_output_cache(self, output_message_cache):
        self.outputMessageCache = output_message_cache

    def terminate(self):
        self.flag = False

    def encode(self):
        while self.flag:

            if self.stand_by:
                time.sleep(1)
                continue

            # send system status to incoming client
            # if self.output_queue.get_length() > 0:
            #     for cpt in range(0, self.output_queue.get_length()):
            #         print "decoding.."
            #         # if self.outputMessageCache.get_latest_guidewire_moving_distance_sequence_length() > 0:
            #         #     msg = self.outputMessageCache.fetch_latest_guidewire_moving_distance_msg()
            #         #     self.output_queue_manager.add_datagram_by_id(cpt, msg)

            time.sleep(self.rtPeriod)
